src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributeList.cs

```diff
@@ -0,0 +1,239 @@
+                        Resources.FormatTagHelperAttributeList_CannotAddWithNullName(
+                            typeof(TagHelperAttribute).FullName,
+                            nameof(TagHelperAttribute.Name)),
+                        nameof(value));
+                }
+
+                Attributes[index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets the first <see cref="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/> matching
+        /// <paramref name="name"/>. When setting, replaces the first matching
+        /// <see cref="TagHelperAttribute"/> with the specified <paramref name="value"/> and removes any additional
+        /// matching <see cref="TagHelperAttribute"/>s. If a matching <see cref="TagHelperAttribute"/> is not found,
+        /// adds the specified <paramref name="value"/> to the end of the collection.
+        /// </summary>
+        /// <param name="name">
+        /// The <see cref="TagHelperAttribute.Name"/> of the <see cref="TagHelperAttribute"/> to get or set.
+        /// </param>
+        /// <returns>The first <see cref="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/> matching
+        /// <paramref name="name"/>.
+        /// </returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively. When setting,
+        /// <see cref="TagHelperAttribute"/>s <see cref="TagHelperAttribute.Name"/> must be <c>null</c> or
+        /// case-insensitively match the specified <paramref name="name"/>.</remarks>
+        /// <example>
+        /// <code>
+        /// var attributes = new TagHelperAttributeList();
+        ///
+        /// // Will "value" be converted to a TagHelperAttribute with a null Name
+        /// attributes["name"] = "value";
+        ///
+        /// // TagHelperAttribute.Name must match the specified name.
+        /// attributes["name"] = new TagHelperAttribute("name", "value");
+        /// </code>
+        /// </example>
+        public new TagHelperAttribute this[[NotNull] string name]
+        {
+            get
+            {
+                return base[name];
+            }
+            [param: NotNull]
+            set
+            {
+                // Name will be null if user attempts to set the attribute via an implicit conversion:
+                // output.Attributes["someName"] = "someValue"
+                if (value.Name == null)
+                {
+                    value.Name = name;
+                }
+                else if (!NameEquals(name, value))
+                {
+                    throw new ArgumentException(
+                        Resources.FormatTagHelperAttributeList_CannotAddAttribute(
+                            nameof(TagHelperAttribute),
+                            nameof(TagHelperAttribute.Name),
+                            value.Name,
+                            name),
+                        nameof(name));
+                }
```

Comment:
looks great but I don't see tests which will hit this `Exception` when the `name` already exists in the list. need data for `StringIndexer_Setter_ThrowsIfUnmatchingKey` where "something" is an existing key.